File: Data/BFNChemistryCatalog.swift
import Foundation

/// Full table of contents — *Big Fat Notebook: High School Chemistry* (ISBN 9781523504251).
enum BFNChemistryCatalog {
    static let editionTitle = "Big Fat Notebook: High School Chemistry"
    static let bookCode = "BFN-C"
    static let isbn = "9781523504251"
    static let chapterCount = 36

    struct Unit: Hashable, Identifiable {
        let number: Int
        let name: String
        let startPage: Int

        var id: Int { number }
    }

    struct Chapter: Hashable, Identifiable {
        let number: Int
        let title: String
        let unitNumber: Int

        var id: Int { number }

        var trackableId: String { "BFN-C-ch-\(number)" }

        var label: String { "Ch \(number) — \(title)" }

        var shortTitle: String { title }
    }

    static let units: [Unit] = [
        Unit(number: 1, name: "Basics of Chemistry", startPage: 1),
        Unit(number: 2, name: "All About Matter", startPage: 73),
        Unit(number: 3, name: "Atomic Theory and Electron Configuration", startPage: 113),
        Unit(number: 4, name: "Elements and the Periodic Table", startPage: 135),
        Unit(number: 5, name: "Bonding and VSEPR Theory", startPage: 179),
        Unit(number: 6, name: "Chemical Compounds", startPage: 231),
        Unit(number: 7, name: "Chemical Reactions and Calculations", startPage: 273),
        Unit(number: 8, name: "Gases", startPage: 311),
        Unit(number: 9, name: "Solutions and Solubility", startPage: 347),
        Unit(number: 10, name: "Acids and Bases", startPage: 383),
        Unit(number: 11, name: "Chemical Equilibrium", startPage: 423),
        Unit(number: 12, name: "Thermodynamics", startPage: 451),
    ]

    static let chapters: [Chapter] = [
        Chapter(number: 1, title: "Introduction to Chemistry", unitNumber: 1),
        Chapter(number: 2, title: "Conducting Experiments", unitNumber: 1),
        Chapter(number: 3, title: "Lab Reports and Evaluating Results", unitNumber: 1),
        Chapter(number: 4, title: "Measurement", unitNumber: 1),
        Chapter(number: 5, title: "Lab Safety and Scientific Tools", unitNumber: 1),
        Chapter(number: 6, title: "Properties of Matter and Changes in Form", unitNumber: 2),
        Chapter(number: 7, title: "States of Matter", unitNumber: 2),
        Chapter(number: 8, title: "Atoms, Elements, Compounds, and Mixtures", unitNumber: 2),
        Chapter(number: 9, title: "Atomic Theory", unitNumber: 3),
        Chapter(number: 10, title: "Waves, Quantum Theory, and Photons", unitNumber: 3),
        Chapter(number: 11, title: "The Periodic Table", unitNumber: 4),
        Chapter(number: 12, title: "Periodic Trends", unitNumber: 4),
        Chapter(number: 13, title: "Electrons", unitNumber: 4),
        Chapter(number: 14, title: "Bonding", unitNumber: 5),
        Chapter(number: 15, title: "Valence Shell Electron Pair Repulsion (VSEPR) Theory", unitNumber: 5),
        Chapter(number: 16, title: "Metallic Bonds and Intramolecular Forces", unitNumber: 5),
        Chapter(number: 17, title: "Naming Substances", unitNumber: 6),
        Chapter(number: 18, title: "The Mole", unitNumber: 6),
        Chapter(number: 19, title: "Finding Compositions in Compounds", unitNumber: 6),
        Chapter(number: 20, title: "Chemical Reactions", unitNumber: 7),
        Chapter(number: 21, title: "Chemical Calculations", unitNumber: 7),
        Chapter(number: 22, title: "Common Gases", unitNumber: 8),
        Chapter(number: 23, title: "Kinetic Molecular Theory", unitNumber: 8),
        Chapter(number: 24, title: "Gas Laws", unitNumber: 8),
        Chapter(number: 25, title: "Solubility", unitNumber: 9),
        Chapter(number: 26, title: "Solubility Rules and Conditions", unitNumber: 9),
        Chapter(number: 27, title: "Concentrations of Solutions", unitNumber: 9),
        Chapter(number: 28, title: "Properties of Acids and Bases", unitNumber: 10),
        Chapter(number: 29, title: "pH Scale and Calculations", unitNumber: 10),
        Chapter(number: 30, title: "Conjugate Acids and Bases", unitNumber: 10),
        Chapter(number: 31, title: "Titrations", unitNumber: 10),
        Chapter(number: 32, title: "Chemical Equilibrium", unitNumber: 11),
        Chapter(number: 33, title: "Le Châtelier's Principle", unitNumber: 11),
        Chapter(number: 34, title: "The First Law of Thermodynamics", unitNumber: 12),
        Chapter(number: 35, title: "The Second Law of Thermodynamics", unitNumber: 12),
        Chapter(number: 36, title: "Reaction Rates", unitNumber: 12),
    ]

    private static let chaptersByNumber: [Int: Chapter] = {
        Dictionary(uniqueKeysWithValues: chapters.map { ($0.number, $0) })
    }()

    private static let unitsByNumber: [Int: Unit] = {
        Dictionary(uniqueKeysWithValues: units.map { ($0.number, $0) })
    }()

    static func chapter(_ number: Int) -> Chapter? {
        chaptersByNumber[number]
    }

    static func unit(_ number: Int) -> Unit? {
        unitsByNumber[number]
    }

    static func unit(forChapter number: Int) -> Unit? {
        chapter(number).flatMap { unit($0.unitNumber) }
    }

    static func pageLabel(forChapter number: Int) -> String {
        guard let unit = unit(forChapter: number) else { return "p1" }
        return "p\(unit.startPage)"
    }

    static func chaptersGroupedByUnit() -> [(unit: Unit, chapters: [Chapter])] {
        units.map { unit in
            (unit, chapters.filter { $0.unitNumber == unit.number })
        }
    }

    static func displayTitle(chapterNumbers: [Int], reviewLabel: String?) -> String {
        if let reviewLabel {
            return reviewLabel
        }
        guard !chapterNumbers.isEmpty else { return "BFN-C reading" }
        let titles = chapterNumbers.compactMap { chapter($0)?.shortTitle }
        if chapterNumbers.count == 1, let n = chapterNumbers.first {
            return "Ch \(n) — \(titles.first ?? "Reading")"
        }
        let range = chapterRangeLabel(chapterNumbers)
        let joined = titles.prefix(2).joined(separator: " · ")
        let suffix = titles.count > 2 ? " · …" : ""
        return "Ch \(range) — \(joined)\(suffix)"
    }

    static func citationLine(chapterNumbers: [Int], reviewLabel: String?) -> String {
        if let reviewLabel {
            return reviewLabel
        }
        guard let first = chapterNumbers.first else {
            return "See index · Match today's chemistry topic"
        }
        let unit = unit(forChapter: first)!
        let chLabel = chapterNumbers.count == 1
            ? "Ch \(first) \(chapter(first)?.title ?? "")"
            : "Ch \(chapterRangeLabel(chapterNumbers))"
        return "Unit \(unit.number) · \(chLabel) · \(pageLabel(forChapter: first))"
    }

    static func optionText(chapterNumbers: [Int], reviewLabel: String?) -> String {
        "\(bookCode) — \(editionTitle) · \(citationLine(chapterNumbers: chapterNumbers, reviewLabel: reviewLabel))"
    }

    static let chapterSectionNames = ["Key Ideas", "Definitions", "Examples", "Mnemonic Devices", "Quiz Yourself"]

    static let chapterSectionGuide = "Key Ideas → Definitions → Examples"

    static func readingLine(chapterNumber: Int) -> String {
        guard let ch = chapter(chapterNumber), let unit = unit(forChapter: chapterNumber) else {
            return "Ch \(chapterNumber)"
        }
        return "Unit \(unit.number) · \(unit.name) · Ch \(ch.number) — \(ch.title) · \(pageLabel(forChapter: chapterNumber))"
    }

    static func compactReadingLine(chapterNumber: Int) -> String {
        guard let ch = chapter(chapterNumber), let unit = unit(forChapter: chapterNumber) else {
            return "Ch \(chapterNumber)"
        }
        return "U\(unit.number) · Ch \(ch.number) — \(ch.shortTitle) · \(pageLabel(forChapter: chapterNumber))"
    }

    static func groupedChapters(_ numbers: [Int]) -> [(unit: Unit, chapters: [Chapter])] {
        var byUnit: [Int: [Chapter]] = [:]
        for number in numbers.sorted() {
            if let chapter = chapter(number) {
                byUnit[chapter.unitNumber, default: []].append(chapter)
            }
        }
        return units.compactMap { unit in
            guard let chapters = byUnit[unit.number], !chapters.isEmpty else { return nil }
            return (unit, chapters)
        }
    }

    private static func chapterRangeLabel(_ numbers: [Int]) -> String {
        guard let first = numbers.first, let last = numbers.last else { return "" }
        if first == last { return "\(first)" }
        return "\(first)–\(last)"
    }
}

File: Scripts/bfn_chemistry_catalog.py
"""BFN-C full book catalog + summer schedule — mirrors Data/BFNChemistrySchedule.swift."""

from __future__ import annotations

from dataclasses import dataclass

UNITS: list[tuple[int, str, int]] = [
    (1, "Basics of Chemistry", 1),
    (2, "All About Matter", 73),
    (3, "Atomic Theory and Electron Configuration", 113),
    (4, "Elements and the Periodic Table", 135),
    (5, "Bonding and VSEPR Theory", 179),
    (6, "Chemical Compounds", 231),
    (7, "Chemical Reactions and Calculations", 273),
    (8, "Gases", 311),
    (9, "Solutions and Solubility", 347),
    (10, "Acids and Bases", 383),
    (11, "Chemical Equilibrium", 423),
    (12, "Thermodynamics", 451),
]

CHAPTERS: list[tuple[int, str, int]] = [
    (1, "Introduction to Chemistry", 1),
    (2, "Conducting Experiments", 1),
    (3, "Lab Reports and Evaluating Results", 1),
    (4, "Measurement", 1),
    (5, "Lab Safety and Scientific Tools", 1),
    (6, "Properties of Matter and Changes in Form", 2),
    (7, "States of Matter", 2),
    (8, "Atoms, Elements, Compounds, and Mixtures", 2),
    (9, "Atomic Theory", 3),
    (10, "Waves, Quantum Theory, and Photons", 3),
    (11, "The Periodic Table", 4),
    (12, "Periodic Trends", 4),
    (13, "Electrons", 4),
    (14, "Bonding", 5),
    (15, "Valence Shell Electron Pair Repulsion (VSEPR) Theory", 5),
    (16, "Metallic Bonds and Intramolecular Forces", 5),
    (17, "Naming Substances", 6),
    (18, "The Mole", 6),
    (19, "Finding Compositions in Compounds", 6),
    (20, "Chemical Reactions", 7),
    (21, "Chemical Calculations", 7),
    (22, "Common Gases", 8),
    (23, "Kinetic Molecular Theory", 8),
    (24, "Gas Laws", 8),
    (25, "Solubility", 9),
    (26, "Solubility Rules and Conditions", 9),
    (27, "Concentrations of Solutions", 9),
    (28, "Properties of Acids and Bases", 10),
    (29, "pH Scale and Calculations", 10),
    (30, "Conjugate Acids and Bases", 10),
    (31, "Titrations", 10),
    (32, "Chemical Equilibrium", 11),
    (33, "Le Châtelier's Principle", 11),
    (34, "The First Law of Thermodynamics", 12),
    (35, "The Second Law of Thermodynamics", 12),
    (36, "Reaction Rates", 12),
]

CHAPTER_BY_NUMBER: dict[int, tuple[str, int]] = {
    number: (title, unit) for number, title, unit in CHAPTERS
}

UNIT_BY_NUMBER: dict[int, tuple[str, int]] = {
    number: (name, start_page) for number, name, start_page in UNITS
}

BOOK_CODE = "BFN-C"
EDITION_TITLE = "Big Fat Notebook: High School Chemistry"


@dataclass(frozen=True)
class DayAssignment:
    chapter_numbers: tuple[int, ...] = ()
    review_label: str | None = None

    @property
    def is_review_day(self) -> bool:
        return self.review_label is not None


def _chapter_range_label(numbers: tuple[int, ...]) -> str:
    if not numbers:
        return ""
    first, last = numbers[0], numbers[-1]
    return str(first) if first == last else f"{first}–{last}"


def _page_label(chapter_number: int) -> str:
    unit_number = CHAPTER_BY_NUMBER[chapter_number][1]
    return f"p{UNIT_BY_NUMBER[unit_number][1]}"


def citation_line(chapter_numbers: tuple[int, ...], review_label: str | None) -> str:
    if review_label:
        return review_label
    if not chapter_numbers:
        return "See index · Match today's chemistry topic"
    first = chapter_numbers[0]
    unit_number = CHAPTER_BY_NUMBER[first][1]
    if len(chapter_numbers) == 1:
        ch_label = f"Ch {first} {CHAPTER_BY_NUMBER[first][0]}"
    else:
        ch_label = f"Ch {_chapter_range_label(chapter_numbers)}"
    return f"Unit {unit_number} · {ch_label} · {_page_label(first)}"


def option_text(chapter_numbers: tuple[int, ...], review_label: str | None = None) -> str:
    return f"{BOOK_CODE} — {EDITION_TITLE} · {citation_line(chapter_numbers, review_label)}"


def _day(chapters: tuple[int, ...] = (), review: str | None = None) -> DayAssignment:
    return DayAssignment(chapter_numbers=chapters, review_label=review)


PLAN: dict[tuple[int, int], DayAssignment] = {
    (1, 0): _day(chapters=(8, 9)),
    (1, 3): _day(chapters=(11, 12)),
    (2, 0): _day(chapters=(9, 10)),
    (2, 3): _day(chapters=(14,)),
    (3, 0): _day(chapters=(25, 26)),
    (3, 3): _day(chapters=(20,)),
    (4, 0): _day(chapters=(6, 7)),
    (4, 3): _day(chapters=(4, 5)),
    (5, 0): _day(chapters=(14, 15)),
    (5, 3): _day(chapters=(18, 19)),
    (6, 0): _day(chapters=(27,)),
    (6, 3): _day(chapters=(28, 29, 31)),
    (7, 0): _day(chapters=(20, 21)),
    (7, 3): _day(chapters=(12, 13)),
    (8, 0): _day(chapters=(4, 19)),
    (8, 3): _day(chapters=(6, 7)),
    (9, 0): _day(chapters=(9, 10)),
    (9, 3): _day(chapters=(16, 17)),
    (10, 0): _day(review="Units 1–6 · Check Your Knowledge"),
    (10, 3): _day(review="Units 7–10 · Check Your Knowledge"),
}


def day_assignment(week: int, day_idx: int) -> DayAssignment | None:
    return PLAN.get((week, day_idx))


def primary_display(week: int, day_idx: int) -> str:
    assignment = day_assignment(week, day_idx)
    if assignment is None:
        return "BFN-C"
    return option_text(assignment.chapter_numbers, assignment.review_label)


def compact_primary_display(week: int, day_idx: int) -> str:
    """Short line for HTML cells: BFN-C — Ch 8–9 — Atoms…"""
    assignment = day_assignment(week, day_idx)
    if assignment is None:
        return "BFN-C"
    if assignment.review_label:
        return f"BFN-C — {assignment.review_label}"
    numbers = assignment.chapter_numbers
    if not numbers:
        return "BFN-C"
    titles = [CHAPTER_BY_NUMBER[n][0] for n in numbers]
    if len(numbers) == 1:
        return f"BFN-C — Ch {numbers[0]} — {titles[0]}"
    joined = " · ".join(titles[:2])
    suffix = " · …" if len(titles) > 2 else ""
    return f"BFN-C — Ch {_chapter_range_label(numbers)} — {joined}{suffix}"


# Hewitt (Expl) § backup — same mapping as BlockAssignedReadingCatalog explChemByKey
HEWITT_CHEM_BACKUP: dict[tuple[int, int], str] = {
    (1, 0): "Ch 17 §17.1–17.3",
    (1, 3): "Ch 17 §17.6–17.8",
    (2, 0): "Ch 15 §15.1–15.5",
    (2, 3): "Ch 18 §18.1–18.4",
    (3, 0): "Ch 19 §19.1–19.4",
    (3, 3): "Ch 20 §20.1–20.4",
    (4, 0): "Ch 17 §17.3–17.5",
    (4, 3): "Ch 17 §17.1–17.4",
    (5, 0): "Ch 18 §18.5–18.8",
    (5, 3): "Ch 19 §19.3–19.4",
    (6, 0): "Ch 19 §19.3–19.5",
    (6, 3): "Ch 21 §21.1–21.3",
    (7, 0): "Ch 20 §20.1–20.4",
    (7, 3): "Ch 17 §17.6–17.7",
    (8, 0): "Ch 17 §17.4–17.5",
    (8, 3): "Ch 17 §17.3–17.5",
    (9, 0): "Ch 17 §17.2–17.3",
    (9, 3): "Ch 18 §18.2–18.4",
    (10, 0): "Review",
    (10, 3): "Review",
}
